### ISSUE TITLE
Why output request body in OpenAPI write operations

### ISSUE THREAD

**Issue by Leo450** [api-platform/api-platform#1891]
**API Platform version(s) affected**: 2.6.4

**Description**  
First of all i'm not sure if that's a bug or a behavior that I do not expect and i think could be modified.

On write operations (POST, PUT, PATCH), OpenAPI documentation uses the output schema (from ```normalizationContext``` groups) to generate the expected request body instead of the input one (from ```denormalizationContext``` groups).

**How to reproduce**  
- A simple Entity where only ```prop2``` should be writable.
```php
namespace App\Domain\MyDomain\Entity;

use ApiPlatform\Core\Annotation\ApiResource;
use App\Domain\Qzd\Repository\MyEntityRepository;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Serializer\Annotation\Groups;

/**
 * @ORM\Entity(repositoryClass=MyEntityRepository::class)
 */
#[ApiResource(
    normalizationContext: ['groups' => ['my_entity:read']],
    denormalizationContext: ['groups' => ['my_entity:write']]
)]
class MyEntity
{
    // ...

    /**
     * @ORM\Column(type="string", length=255)
     */
    #[Groups(['my_entity:read'])]
    private $prop1;

    /**
     * @ORM\Column(type="string", length=255)
     */
    #[Groups(['my_entity:write'])]
    private $prop2;

    // ...
}
```

- What i have in OpenAPI request body for the ```POST``` operation

![bug_openapi](https://user-images.githubusercontent.com/7643118/115795013-5f751b00-a3cf-11eb-92df-d6bac2653210.png)

I'm kinda new to API Platform so maybe there is a misunderstanding on my side but yeah, is this the expected behavior or some kind of bug ?

I have written an ```OpenApiFactory``` decorator to solve this issue if someone's interested.

Thanks for your time and concern.

**Comment by soullivaneuh** [comment]
> I have written an OpenApiFactory decorator to solve this issue if someone's interested.

@Leo450 I am. :smile:

**Comment by Leo450** [comment]
> > I have written an OpenApiFactory decorator to solve this issue if someone's interested.
> 
> @Leo450 I am.

Here it is.
I wrote this fast, can be improved for sure and if you do improve it, i would be glad to look at your code.
```php
<?php

namespace App\Infrastructure\OpenApi;

use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
use ApiPlatform\Core\OpenApi\Factory\OpenApiFactoryInterface;
use ApiPlatform\Core\OpenApi\Model;
use ApiPlatform\Core\OpenApi\Model\PathItem;
use ApiPlatform\Core\OpenApi\OpenApi;
use Doctrine\ORM\EntityManagerInterface;

class MyOpenApiFactory implements OpenApiFactoryInterface
{
    private array $entities = [];
    private array $schemas = [];
    private array $schemaKeys = [];

    public function __construct(
        private OpenApiFactoryInterface $decorated,
        private ResourceMetadataFactoryInterface $metadataFactory,
        EntityManagerInterface $entityManager
    )
    {
        // List all registered entities indexed by short name
        foreach ($entityManager->getConfiguration()->getMetadataDriverImpl()->getAllClassNames() as $entity) {
            $path = explode('\\', $entity);
            $this->entities[array_pop($path)] = $entity;
        }
    }

    public function __invoke(array $context = []): OpenApi
    {
        $openApi = $this->decorated->__invoke($context);

        $this->schemas = $openApi->getComponents()->getSchemas()->getArrayCopy();
        $this->schemaKeys = array_keys($this->schemas);

        $finalPaths = new Model\Paths();
        /** @var Model\PathItem $pathItem */
        foreach ($openApi->getPaths()->getPaths() as $path => $pathItem) {
            $this->handleWriteOperationsRequestBody($path, $pathItem);
            $finalPaths->addPath($path, $pathItem);
        }

        return $openApi->withPaths($finalPaths);
    }

    private function handleWriteOperationsRequestBody(string $path, PathItem &$pathItem)
    {
        $this->handleWriteOperationRequestBody($path, $pathItem, 'post');
        $this->handleWriteOperationRequestBody($path, $pathItem, 'put');
        $this->handleWriteOperationRequestBody($path, $pathItem, 'patch');
    }

    /**
     * @bug On write operations (POST, PUT, PATCH), OpenAPI documentation uses the output schema (from normalizationContext groups) to generate the expected request body instead of the input one (from denormalizationContext groups).
     * @see https://github.com/api-platform/api-platform/issues/1891
     *
     * Change write operations request body to use the input schema instead of the output one
     */
    private function handleWriteOperationRequestBody(string $path, PathItem &$pathItem, string $operationMethod)
    {
        $operationMethod = strtoupper($operationMethod);
        $operationName = strtolower($operationMethod);
        $operationGetter = 'get' . ucfirst(strtolower($operationMethod));
        $operationWither = 'with' . ucfirst(strtolower($operationMethod));
        $operationsGetter = $operationMethod === 'POST' ? 'getCollectionOperations' : 'getItemOperations';
        $operation = $pathItem->$operationGetter();

        if(!$operation)
            return;

        /** ------------------------------------
         * Resolve entity class
         * ------------------------------------- */
        $entityClass = null;
        foreach ($operation->getTags() as $tag) {
            if(array_key_exists($tag, $this->entities)){
                $entityClass = $this->entities[$tag];
                break;
            }
        }
        if(!$entityClass)
            return;

        /** ------------------------------------
         * Resolve entity resource metadata
         * ------------------------------------- */
        $entityMetadata = $this->metadataFactory->create($entityClass);

        /** ------------------------------------
         * Resolve entity resource metadata
         * ------------------------------------- */
        $operationMetadata = null;
        // Look in collectionOperation for operation with good path and method
        foreach ($entityMetadata->$operationsGetter() as $_operation) {
            if(
                array_key_exists('path', $_operation)
                && $_operation['path'] === $path
                && array_key_exists('method', $_operation)
                && strtoupper($_operation['method']) === $operationMethod
            )
                $operationMetadata = $_operation;
        }
        // If not found, look for operation with good name
        if(!$operationMetadata){
            foreach ($entityMetadata->$operationsGetter() as $_operationName => $_operation) {
                if($_operationName === $operationName)
                    $operationMetadata = $_operation;
            }
        }
        if(!$operationMetadata)
            return;

        /** ------------------------------------
         * Resolve operation denormalization context
         * ------------------------------------- */
        $denormalizationContext = null;
        // Look for an operation level definition
        if(array_key_exists('denormalization_context', $operationMetadata)){
            $denormalizationContext = $operationMetadata['denormalization_context'];
        }
        // If not found, look for a resource level definition
        if(!$denormalizationContext){
            $attributes = $entityMetadata->getAttributes();
            if(array_key_exists('denormalization_context', $attributes) && $attributes['denormalization_context']){
                $denormalizationContext = $attributes['denormalization_context'];
            }
        }
        if(!$denormalizationContext)
            return;

        /** ------------------------------------
         * Resolve new schema
         * ------------------------------------- */
        $schemaRef = str_replace(':', '.', "{$entityMetadata->getShortName()}-" . implode('_', $denormalizationContext['groups']));
        if(!in_array($schemaRef, $this->schemaKeys))
            return;
        $schema = $this->schemas[$schemaRef];

        /** ------------------------------------
         * Apply new schema
         * ------------------------------------- */
        $requestBody = $operation->getRequestBody();
        $contents = $requestBody->getContent();
        foreach ($contents as $type => $content){
            if(is_array($content))
                continue;
            $contents[$type] = $content->withSchema($schema);
        }
        $requestBody = $requestBody->withContent($contents);
        $operation = $operation->withRequestBody($requestBody);
        $pathItem = $pathItem->$operationWither($operation);
    }
}
```

**Comment by tuancode** [comment]
I got the same issue after upgrade to version 2.6.4. The OpenAPI cannot detect denormalization context group to render properly request body for a write operation. It instead render the nomarlization group context.

**Comment by alanpoulain** [comment]
Solved in 2.6.5.
